include/deal.II/numerics/smoothness_estimator.templates.h

```diff
@@ -0,0 +1,434 @@
+  void
+  legendre_coeff_decay(
+    FESeries::Legendre<dim, spacedim> &    fe_legendre,
+    const hp::DoFHandler<dim, spacedim> &  dof_handler,
+    const std::vector<const VectorType *> &all_solutions,
+    const std::vector<Vector<float> *> &   all_smoothness_indicators,
+    const double                           smallest_abs_coefficient)
+  {
+    Assert(smallest_abs_coefficient >= 0.,
+           ExcMessage("smallest_abs_coefficient should be non-negative."));
+
+    using number = typename VectorType::value_type;
+    using number_coeff =
+      typename FESeries::Legendre<dim, spacedim>::CoefficientType;
+
+    AssertDimension(all_solutions.size(), all_smoothness_indicators.size());
+
+    for (auto &smoothness_indicator : all_smoothness_indicators)
+      smoothness_indicator->reinit(
+        dof_handler.get_triangulation().n_active_cells());
+
+    Table<dim, number_coeff> expansion_coefficients;
+    resize(expansion_coefficients, fe_legendre.get_size_in_each_direction());
+
+    Vector<number> local_dof_values;
+
+    // auxiliary vector to do linear regression
+    std::vector<number_coeff> x;
+    std::vector<number_coeff> y;
+
+    x.reserve(dof_handler.get_fe_collection().max_degree());
+    y.reserve(dof_handler.get_fe_collection().max_degree());
+
+    for (auto &cell : dof_handler.active_cell_iterators())
+      if (cell->is_locally_owned())
+        {
+          local_dof_values.reinit(cell->get_fe().dofs_per_cell);
+
+          const unsigned int pe = cell->get_fe().degree;
+
+          // since we use coefficients with indices [1,pe] in each direction,
+          // the number of coefficients we need to calculate is at least N=pe+1
+          AssertIndexRange(pe, fe_legendre.get_size_in_each_direction());
+
+          auto solution_it              = all_solutions.cbegin();
+          auto smoothness_indicators_it = all_smoothness_indicators.begin();
+          for (; solution_it != all_solutions.cend();
+               ++solution_it, ++smoothness_indicators_it)
+            {
+              // for linear elements set analyticity to 0 (i.e. smooth)
+              // as we can't really estimate anything reasonably
+              if (pe == 1)
+                {
+                  (*(*smoothness_indicators_it))(cell->active_cell_index()) = 0;
+                  continue;
+                }
+
+              cell->get_dof_values(*(*solution_it), local_dof_values);
+              fe_legendre.calculate(local_dof_values,
+                                    cell->active_fe_index(),
+                                    expansion_coefficients);
+
+              // choose the smallest decay of coefficients in each direction,
+              // i.e. the maximum decay slope k_v
+              number_coeff k_v = -std::numeric_limits<number_coeff>::max();
+              for (unsigned int d = 0; d < dim; d++)
+                {
+                  x.resize(0);
+                  y.resize(0);
```

Comment:
I suspect that a `resize(0)` also releases all memory. You may want to call `x.reserve(...)` here and the same for `y`.

Reply:
according to this [stackoverflow](https://stackoverflow.com/a/1155710/888478) `resize` won't change capacity. I do `reserve` above (outside of the cell loop).